test/test_session.py

```diff
@@ -173,6 +175,40 @@ def _test_ops(self, client, *ops):
                         "%s did not return implicit session to pool" % (f.__name__,),
                     )
 
+    def test_implicit_sessions_checkout(self):
+        # "To confirm that implicit sessions only allocate their server session after a
+        # successful connection checkout" test from Driver Sessions Spec.
+        listener = EventListener()
+        client = rs_or_single_client(event_listeners=[listener], maxPoolSize=1, retryWrites=True)
+        cursor = client.db.test.find({})
+        ops: List[Tuple[Callable, List[Any]]] = [
+            (client.db.test.find_one, [{"_id": 1}]),
+            (client.db.test.delete_one, [{}]),
+            (client.db.test.update_one, [{}, {"$set": {"x": 2}}]),
+            (client.db.test.bulk_write, [[UpdateOne({}, {"$set": {"x": 2}})]]),
+            (client.db.test.find_one_and_delete, [{}]),
+            (client.db.test.find_one_and_update, [{}, {"$set": {"x": 1}}]),
+            (client.db.test.find_one_and_replace, [{}, {}]),
+            (client.db.test.aggregate, [[{"$set": {"x": 1}}]]),
+            (client.db.test.find, [{}]),
+            (client.server_info, [{}]),
+            (cursor.distinct, ["_id"]),
+            (client.db.list_collections, [{}]),
+        ]
+        threads = []
+        listener.results.clear()
+        for op, args in ops:
+            threads.append(threading.Thread(target=op, args=args))
+            threads[-1].start()
+        for thread in threads:
+            thread.join()
+
+        lsid_set = set()
+        for i in listener.results["started"]:
+            if i.command.get("lsid"):
+                lsid_set.add(i.command.get("lsid")["id"])
+        self.assertEqual(len(lsid_set), 1)
```

Comment:
I worry a bit that this test will be flaky. The issue is that the session is checked back into the pool _after_ the connection is checked back in. So there's always the risk that this test uses >1 server session. Let's try a "wait and see" approach where we merge the test as is while keeping in mind that it might fail.

Reply:
I actually did notice this test failing in CI one time. So I believe you are correct. However, I tried to replicate it locally by running the test 1000s of times and could not reproduce.

Reply:
Thanks for opening that drivers ticket. Right now it appears that on replica sets it still fails even with 5 iterations.

Reply:
How was the test passing on replica sets before?